(/basic-rbac) add middleware, reorg and refactor copy (#1560)
clerkinc/clerk-docs: docs/guides/basic-rbac.mdx

description: Learn how to leverage Clerk's publicMetadata to implement your own basic Role Based Access Controls.
---

To control which users can access certain parts of your application, you can leverage Clerk's [roles](/docs/organizations/roles-permissions#roles) feature. Although Clerk offers a roles feature as part of the feature set for [organizations](/docs/organizations/overview), not every app implements organizations. **This guide covers a workaround to set up a basic Role Based Access Control (RBAC) system for products that don't use Clerk's organizations or roles.**

> This guide assumes that you're using Next.js App Router, but the concepts can be adapted to any React-based framework.

Reviewer: This is not really true -- this is leveraging server side checks on the role and then redirecting the user from the route. A React app can't do this. To do this on React you would need to leverage an Express or other server side app to handle much of the logic and the implementation would be different. I would revert to the previous statement.

Reviewer 2: [c2ee7a1](https://github.com/clerk/clerk-docs/pull/1560/commits/c2ee7a1f4a008ea6e51f4d097dda612f213575cb)

<Steps>
  ### Configure the session token

  Clerk provides [user metadata](/docs/users/metadata#user-metadata), which can be used to store information, and in this case, it can be used to store a user's role.

  There are three types of metadata:

  - `unsafeMetadata` - can be both read and modified in the browser, meaning users can modify this data. Therefore, you should validate `unsafeMetadata` before trusting it for sensitive operations.
  - `privateMetadata` - cannot be accessed nor modified in the browser.

>   - `publicMetadata` - can be accessed in the browser, but only be modified server-side or in the Clerk Dashboard.

Reviewer: The revisions on this section have been frequent. Maybe we don't repeat this and instead link to the metadata section? That said, this list is accurate.

Reviewer 2: [cf6613c](https://github.com/clerk/clerk-docs/pull/1560/commits/cf6613ce0a01e144d691d4098372d1867080757e)

  Since `publicMetadata` can only be read but not modified in the browser, it is the safest and most appropriate choice for storing information

>   To build a basic RBAC system, you first need to make `publicMetadata` available to the application directly from the session token. By attaching `publicMetadata` to the user's session, you can access the data without needing to make a fetch or network request each time.

Reviewer: > fetch or network request
This isn't really an 'or'. I'd say 'fetch and network request' where fetch describes the function/action and network request refers to the latency. Barring that I would say to use just fetch or just network request. eg

> without needing to make a fetch each time.

Reviewer 2: [6bc6229](https://github.com/clerk/clerk-docs/pull/1560/commits/6bc62299498e9c94b7d9a789f9387098108452c6)

  1. Navigate to the [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=sessions).
  1. In the top navigation, select **Configure**. Then in the sidebar, select **Sessions**.
  1. Under the **Customize session token** section, select **Edit**.
  1. In the modal that opens, enter the following JSON and select **Save**. If you have already customized your session token, you may need to merge this with what you currently have.

  ```json
  {
    "metadata": "{{user.public_metadata}}"
  }
  ```

  > [!CAUTION]
  > The session token has a 4KB size limit. Exceeding this limit can have adverse effects, such as an infinite redirect loop for users in Next.js applications.
  > To avoid this, it's recommended to move large claims out of the JWT and fetch them via a separate API call from your backend.

  ### Create a global TypeScript definition

  1. In your application's root folder, create a `types` directory.

>   1. Inside this directory, add a `globals.d.ts` file. This file will provide auto-completion, prevent TypeScript errors when working with roles, and control which roles are allowed in the application.

Reviewer: ```suggestion
  1. Inside this directory, add a `globals.d.ts` file. This file will provide auto-completion and prevent TypeScript errors when working with roles.
```

> and control which roles are allowed in the application
This isn't really true. If you are using TypeScript and respecting linting then the app won't deploy with the errors but that's less about controlling the roles than alerting you to errors where you specific another role. Also the server action that sets the metadata/role doesn't check its one of the roles in `Roles` cause its a super basic example.

  export const checkRole = (role: Roles) => {
    const { sessionClaims } = auth()
    return sessionClaims?.metadata.role === role
  }
  ```

  > [!NOTE]
  > You can customize the behavior of the `checkRole()` helper function to suit your needs. For example, you could modify it to return the roles a user has or create a `protectByRole()` function that handles role-based redirects.

  ### Create the admin dashboard

  Now, it's time to create an admin dashboard. The first step is to create the `/admin` route.

  1. In your `app/` directory, create an `admin/` folder.
  1. In the `admin/` folder, create a file named `page.tsx`.
  1. Add the following placeholder code to the file.

  ```tsx {{ filename: 'app/admin/page.tsx' }}
  export default function AdminDashboard() {
    return <p>This is the protected admin dashboard restricted to users with the `admin` role.</p>
  }
  ```

  ### Protect the admin dashboard

  To protect the `/admin` route, choose **one** of the two following methods:

  1. **Middleware**: Apply role-based access control globally at the route level. This method restricts access to all routes matching `/admin` before the request reaches the actual page.

>   1. **Page-Level Role Check**: Apply role-based access control directly in the admin page component. This method protects the page itself but doesn't block access at the middleware level.

Reviewer: ```suggestion
  1. **Page-Level Role Check**: Apply role-based access control directly in the admin page component. This method protects the page itself but does. not protect other pages/routes in the admin dashboard. If you also have a `/admin/invoices` route you would need to also apply this protection to that route.
```
That was a quick edit. The bit about Middleware is correct, but I think pointing out the need to protect other pages is more helpful here. That said, reword as you feel best.

Reviewer 2: [b20141c](https://github.com/clerk/clerk-docs/pull/1560/commits/b20141cd7ae85be7bc56f4a6feb71805fbee63c6)

  ```ts {{ filename: 'app/admin/_actions.ts' }}
  'use server'

  import { checkRole } from '@/utils/roles'
  import { clerkClient } from '@clerk/nextjs/server'

  export async function setRole(formData: FormData) {
    // Check that the user trying to set the role is an admin
    if (!checkRole('admin')) {
      return { message: 'Not Authorized' }
    }

    try {
      const res = await clerkClient().users.updateUser(formData.get('id') as string, {
        publicMetadata: { role: formData.get('role') },
      })
      return { message: res.publicMetadata }
    } catch (err) {
      return { message: err }
    }
  }
  ```

  With the server action set up, now build the `<SearchUsers />` component. This component includes a form for searching users, and when submitted, appends the search term to the URL as a search parameter. Your `/admin` route will then perform a query based on the updated URL.

>   1. In your `app/admin/` directory, create a `_search-users.tsx` file.

Reviewer: Not required, but I would either:
1. Use the name `_components.tsx` as that's a little more common
2. Move the file and use `/components/SearchUsers.tsx` as that's definitely more common.

  1. Add the following code to the file.

  ```tsx {{ filename: 'app/admin/_search-users.tsx' }}
  'use client'

  import { usePathname, useRouter } from 'next/navigation'

  export const SearchUsers = () => {
    const router = useRouter()
    const pathname = usePathname()

    return (
      <div>
        <form
          onSubmit={(e) => {
            e.preventDefault()
            const form = e.currentTarget
            const formData = new FormData(form)
            const queryTerm = formData.get('search') as string
            router.push(pathname + '?search=' + queryTerm)
          }}
        >
          <label htmlFor="search">Search for Users</label>
          <input id="search" name="search" type="text" />
          <button type="submit">Submit</button>
        </form>
      </div>
    )
  }
  ```

>   With the server action and the search form set up, it's time to refactor the `app/admin/page.tsx` to be a server component. It will check whether a search parameter has been appended to the URL by the search form. If a search parameter is present, it will query for users matching the entered term.

Reviewer: ```suggestion
  With the server action and the search form set up, it's time to refactor the `app/admin/page.tsx`. It will check whether a search parameter has been appended to the URL by the search form. If a search parameter is present, it will query for users matching the entered term.
```

The admin page was already a server component.

>   If one or more users are found, the component will display a list of users, showing their first and last names, primary email address, and current role. Each user will have `Make Admin` and `Make Moderator` buttons, which include hidden inputs for the user ID and role. These buttons will use the `setRole()` server action to update the user's role.

Reviewer: It might be worth adding a third button called `Clear Roles`, and a server action shockingly called `clearRoles()` which nulls the role metadata, removing any roles and returning the user to a normal user.

Reviewer 2: [65e009e](https://github.com/clerk/clerk-docs/pull/1560/commits/65e009eba507215c97f8ed5ade542d616b465c91)

Reviewer: The server action looks good, but there is no button in the page.tsx. Something like the following (untested!)

```tsx
import { setRole, clearRole } from './_actions'

<form action={clearRole}>
  <input type="hidden" value={user.id} name="id" />
  <button type="submit">Remove Role</button>
</form>
```

Reviewer: Note, I used the function name as I suggested in the (`removeRole` instead of `removeRoles`). If you don't change the function name, this needs to be adjusted accordingly.

Reviewer 2: oh I had in my code but forgot to add to the doc LOL

    return (
      <>
        <p>This is the protected admin dashboard restricted to users with the `admin` role.</p>

        <SearchUsers />

        {users.map((user) => {
          return (
            <div key={user.id}>
              <div>
                {user.firstName} {user.lastName}
              </div>

              <div>
                {
                  user.emailAddresses.find((email) => email.id === user.primaryEmailAddressId)
                    ?.emailAddress
                }
              </div>

              <div>{user.publicMetadata.role as string}</div>

              <form action={setRole}>
                <input type="hidden" value={user.id} name="id" />
                <input type="hidden" value="admin" name="role" />
                <button type="submit">Make Admin</button>
              </form>

              <form action={setRole}>
                <input type="hidden" value={user.id} name="id" />
                <input type="hidden" value="moderator" name="role" />
                <button type="submit">Make Moderator</button>
              </form>
            </div>
          )
        })}
      </>
    )
  }
  ```

>   ### Finished!

Reviewer: I miss the 🎉 (not a plea to add it, just a silly comment)

Reviewer 2: [dd80c54](https://github.com/clerk/clerk-docs/pull/1560/commits/dd80c5481f292b5631858661dd17fd431588d03e)

hehe

  Create a server action for managing a user's role.

  1. In your `app/admin/` directory, create an `_actions.ts` file.
  1. Create a server action that sets a user's role. Use the `checkRole()` function to verify that the current user has the `admin` role. If they do, proceed to update the specified user's role using Clerk's [JavaScript Backend SDK](/docs/references/backend/user/update-user). This ensures that only administrators can modify user roles.
  1. Create a server action that removes a user's role.

  ```ts {{ filename: 'app/admin/_actions.ts' }}
  'use server'

  import { checkRole } from '@/utils/roles'
  import { clerkClient } from '@clerk/nextjs/server'

  export async function setRole(formData: FormData) {
    // Check that the user trying to set the role is an admin
    if (!checkRole('admin')) {
      return { message: 'Not Authorized' }
    }

    try {
      const res = await clerkClient().users.updateUser(formData.get('id') as string, {
        publicMetadata: { role: formData.get('role') },
      })
      return { message: res.publicMetadata }
    } catch (err) {
      return { message: err }
    }
  }

>   export async function removeRoles(formData: FormData) {

Reviewer: nit: I would use `removeRole` since this is a per user function and a user only has one role. That said, nothing wrong this either.

description: Learn how to leverage Clerk's publicMetadata to implement your own basic Role Based Access Controls.
---

To control which users can access certain parts of your application, you can leverage Clerk's [roles](/docs/organizations/roles-permissions#roles) feature. Although Clerk offers a roles feature as part of the feature set for [organizations](/docs/organizations/overview), not every app implements organizations. This guide covers a workaround to set up a basic Role Based Access Control (RBAC) system for products that don't use Clerk's organizations or roles.

This guide assumes that you're using Next.js App Router. The concepts here can be adapted to Next.js Pages Router and Remix.

## Configure the session token

Clerk provides [user metadata](/docs/users/metadata#user-metadata), a tool that can be leveraged to build flexible custom logic into your application. Metadata can be used to store information, and in this case, it can be used to store a user's role.

> `unsafeMetadata` can be read and updated in the browser. Since users can modify this metadata, it should be considered untrusted and must be validated by your application before use. `privateMetadata` can not be accessed or modified in the browser. `publicMetadata` can be read by the browser and can only be updated server-side or in the Clerk Dashboard, making it the safest and best choice for this use case.

Reviewer: > Since users can modify this metadata, it should be considered untrusted and must be validated by your application before use.

The vagueness of the previous version is actually more accurate. You don't always need trust' t he `unsateMetadata` depending on how you are using it. For example, a user could self-report a job title (think LinkedIn). That can't be validated and the application isn't going to perform a privileged operation based on that. It would just display it.

Reviewer: That looks good.

> `unsafeMetadata` can be read and updated in the browser. Since users can modify this metadata, it should be considered untrusted and must be validated by your application before use. `privateMetadata` can not be accessed or modified in the browser. `publicMetadata` can be read by the browser and can only be updated server-side or in the Clerk Dashboard, making it the safest and best choice for this use case.

Reviewer: The only part you might want to drop is the `In contrast, `, as that reads, at least to me, like there are just two options when there are actually 3.

## Set the admin role for your user

Later in the, you will add a basic admin tool to change a user's role. For now, let's manually add a role to your own user account.

1. In the Clerk Dashboard, navigate to [**Users**](https://dashboard.clerk.com/last-active?path=users) and select your own user account.
1. Scroll down to the **User metadata** section and next to the **Public** option, select **Edit**.
1. Add the following JSON and select **Save**.

```json
{
  "role": "admin"
}
```

## Create and protect an admin dashboard

Now that your user has the `admin` role, let's build an admin dashboard manage user roles easily.

1. In your `app/` directory, create an `admin/` folder.

> 1. Inside the `admin/` folder, create a `dashboard/` folder

Reviewer: ```suggestion
```

The `/admin/dashboard` double layer is confusing. I don't recall how that got into the original guide, but there its not reflective of a feature like this. In the below suggestions I am changing it to just `/admin` though just `/dashboard` would also work.

> 1. In the `dashboard/` folder, create a file named `page.tsx`.

Reviewer: ```suggestion
1. In the `admin/` folder, create a file named `page.tsx`.
```

1. Copy the following code and paste it into the file.

You want the dashboard to be accessible only to users with the `admin` role. As configured earlier in the [Configure the session token](#configure-the-session-token) section, a user's role is stored in the metadata within the session token. You can access the session token's claims using Clerk's `auth()` hook.

> ```tsx {{ filename: 'app/admin/dashboard/page.tsx' }}

Reviewer: ```suggestion
```tsx {{ filename: 'app/admin/page.tsx' }}
```

import { auth } from '@clerk/nextjs/server'
import { redirect } from 'next/navigation'

> export default async function AdminDashboard() {

Reviewer: ```suggestion
export default function AdminDashboard() {
```
There is nothing that requires an `await` so this shouldn't be swapped to an `async` function.

  const { sessionClaims } = auth()

  // If the user does not have the admin role, redirect them to the home page
  if (sessionClaims?.metadata.role !== 'admin') {
    redirect('/')
  }

  return <p>This is the protected admin dashboard restricted to users with the `admin` role.</p>
}
```

> The `/admin/dashboard` route now requires the user to sign in to access the application. Additionally, users must have a `publicMetadata` value of `{"role": "admin"}` to access this route.

Reviewer: ```suggestion
The `/admin` route now requires the user to sign in to access the application. Additionally, users must have a `publicMetadata` value of `{"role": "admin"}` to access this route.
```

}
```

The next step is to create the helper function for checking roles.

1. Create a `utils/` directory in your application's root folder.
1. Inside this directory, add a `roles.ts` file.
1. Add the following code to `roles.ts`.

```ts {{ filename: 'utils/roles.ts' }}
import { Roles } from '../types/globals'
import { auth } from '@clerk/nextjs/server'

export const checkRole = (role: Roles) => {
  const { sessionClaims } = auth()
  return sessionClaims?.metadata.role === role
}
```

This `checkRole()` helper will accept a role of type `Roles` and return `true` if the user has that role, or `false` if the they do not.

Next, refactor the admin dashboard to use the `checkRole()` helper.

1. Navigate back to the admin dashboard file.
1. In the `if()` statement, remove the previous role check.
1. Replace it with the new `checkRole()` helper, passing `"admin"` as the argument.

> ```tsx {{ filename: 'app/admin/dashboard/page.tsx', mark: [2, [5, 8]] }}

Reviewer: ```suggestion
```tsx {{ filename: 'app/admin/page.tsx', mark: [2, [5, 8]] }}
```

import { redirect } from 'next/navigation'
import { checkRole } from '@/utils/roles'

export default async function AdminDashboard() {
  // If the user does not have the admin role, redirect them to the home page
  if (!checkRole('admin')) {
    redirect('/')
  }

  return <p>This is the protected admin dashboard restricted to users with the `admin` role.</p>
}
```

> [!NOTE]
> You can customize the behavior of the `checkRole()` helper function to suit your needs. For example, you could modify it to return the roles a user has or create a `protectByRole()` function that handles role-based redirects.

## Add admin tools to find users and add roles

You can use the `checkRole()` function along with server actions to build basic tools for finding users and managing roles.

Start with the server action. The following `setRole()` action will first check the role of the current user using the `checkRoles()` helper to confirm that they are an `admin`. It will then set the specified role for the selected user.

> ```ts {{ filename: 'app/admin/dashboard/_actions.ts' }}

Reviewer: ```suggestion
```ts {{ filename: 'app/admin/_actions.ts' }}
```

'use server'

> import { checkRole } from '../../../utils/roles'

Reviewer: Next by default uses  the `@` for an alias import. I feel like that's a reasonable usage in our docs. If we aren't going to use that, then I think an absolute part (`app/utils/roles`) is probably more clear than a relative path.

}
```

With the server action in place, you can now build the `<SearchUsers />` component. This component will include a form for searching users. When the form is submitted, the search term is appended to the URL as a search parameter. The page component, which is a server component that you will refactor next, will perform a query based on the updated URL.

> ```tsx {{ filename: 'app/admin/dashboard/_search-users.tsx' }}

Reviewer: ```suggestion
```tsx {{ filename: 'app/admin/_search-users.tsx' }}
```

}
```

> With the server action and the search form in place, you'll now refactor the server component for the `app/admin/dashboard` route. The component will check if a search parameter has been added to the URL by the search form. If a search parameter is present, it will search for users matching the entered term.

Reviewer: ```suggestion
With the server action and the search form in place, you'll now refactor the server component for the `app/admin` route. The component will check if a search parameter has been added to the URL by the search form. If a search parameter is present, it will search for users matching the entered term.
```

If an array of one or more users is returned, then the component will display a list of users including their first and last names, primary email address, and current role. Each user will have `Make Admin` and `Make Moderator` buttons. These buttons include hidden inputs for the user ID and role, and they use the `setRole()` server action to update the user's role.

> ```tsx {{ filename: 'app/admin/dashboard/page.tsx', mark: [[3, 5], 7, 12, 14, 17, [20, 52]] }}

Reviewer: ```suggestion
```tsx {{ filename: 'app/admin/page.tsx', mark: [[3, 5], 7, 12, 14, 17, [20, 52]] }}
```
